server/connmux/connmux_test.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connmux
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+	"time"
+
+	"net/http"
+	"net/url"
+	"testing"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	testpb "google.golang.org/grpc/test/grpc_testing"
+)
+
+func TestConnMux(t *testing.T) {
+	certFile := "../../tests/fixtures/server.crt"
+	keyFile := "../../tests/fixtures/server.key.insecure"
+
+	tcpListener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("error creating listener: %v", err)
+	}
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		t.Errorf("failed to load TLS keys: %v", err)
+	}
+	connMux := New(Config{
+		Logger:   zap.NewExample(),
+		Listener: tcpListener,
+		Insecure: true,
+		Secure:   true,
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			MinVersion:   tls.VersionTLS12,
+		},
+	})
+
+	t.Logf("Listening on address %s", connMux.root.Addr().String())
+
+	// HTTP server
+	httpL := connMux.HTTPListener()
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "HTTP response Proto: %s %v\n", r.Proto, r.Header)
+	})
+	httpServer := &http.Server{Handler: mux}
+	go httpServer.Serve(httpL)
+
+	// GRPC Server
+	grpcServer := grpc.NewServer()
+	testpb.RegisterTestServiceServer(grpcServer, &dummyStubServer{body: []byte("grpc response")})
+
+	grpcL := connMux.GRPCListener()
+	go grpcServer.Serve(grpcL)
+
+	go connMux.Serve()
+	defer connMux.Close()
+
+	destination := url.URL{Scheme: "https", Host: connMux.root.Addr().String()}
+
+	// Set up a connection to the server.
+	conn, err := grpc.Dial(destination.Host, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+
+	for i := 0; i < 10; i++ {
```

Comment:
why do you need 10 repetitions here? 

I'm also not sure if it makes sense to have a parameterized and separated test run for each type so it's more visible what fails in CI.

Reply:
gotcha, it's not really doing any concurrent calls as far as I can see?